### PR TITLE
[release/3.1] Fix build failures with clang 13 rc1

### PR DESCRIPTION
**Pull request by omajid**
Clang 13 now produces an error in json/casablanca because a function
argument is not used:

    In file included from /core-setup/src/corehost/cli/deps_format.cpp:6:
    In file included from /core-setup/src/corehost/cli/deps_format.h:14:
    In file included from /core-setup/src/corehost/cli/fxr/../json/casablanca/include/cpprest/json.h:36:
    In file included from /core-setup/src/corehost/cli/fxr/../json/casablanca/include/cpprest/details/basic_types.h:41:
    /core-setup/src/corehost/cli/fxr/../json/casablanca/include/cpprest/details/SafeInt3.hpp:1317:33: error: parameter 'b' set but not used [-Werror,-Wunused-but-set-parameter]
        static void CastThrow( bool b, T& t ) SAFEINT_CPP_THROW
                                    ^
    1 error generated.
    gmake[2]: *** [cli/fxr/CMakeFiles/hostfxr.dir/build.make:76: cli/fxr/CMakeFiles/hostfxr.dir/__/deps_format.cpp.o] Error 1

Cast the argument to void to avoid this warning/error.

### CODE REVIEW

**Comment by Anipik** [comment]
cc @NikolaMilosavljevic can you review this change

**Comment by omajid** [comment]
Ping. Anyone want to review this?

**Comment by Anipik** [comment]
cc @jeffschwMSFT @aik-jahoda

**Comment by jeffschwMSFT** [comment]
Adding @janvorli and @mangod9

**Comment by mangod9** [comment]
The change itself looks innocuous, but I am not sure about whether updating to clang 13 is required for rc1?

**Comment by janvorli** [comment]
@mangod9 the "rc1" refers to clang 13 rc1, not .NET rc1.

**Comment by safern** [comment]
Is this change still needed?

**Comment by omajid** [comment]
It would be nice to include this to support building against clang 13. That said, this change is now known to be incomplete; we will need the changes from https://github.com/dotnet/runtime/pull/63314 as well.

**Comment by safern** [comment]
I see. Should we then close this PR until it can be worked on to get it to a complete state?

**Comment by omajid** [comment]
I converetd the PR to draft, let me know if it's still better to close it until I get a chance to re-visit it.

**Comment by safern** [comment]
That's good. Thanks, it just helps to have only open PRs on these repos that are targeting the next servicing release.

**Comment by omajid** [comment]
Uh.. I opened https://github.com/dotnet/core-setup/pull/9197 yesterday which fixes the same issue differently. Any thoughts which approach is better?

**Comment by omajid** [comment]
Any thought about this?

**Comment by NikolaMilosavljevic** [comment]
@omajid where do you see this error? I presume this is a local dev build, as we do not get this error in official builds.

**Comment by NikolaMilosavljevic** [comment]
cc @vitek-karas

**Comment by omajid** [comment]
@NikolaMilosavljevic I am one of the maintainer of [the .NET Core 3.1 package in Fedora](https://src.fedoraproject.org/rpms/dotnet3.1/tree/f36). Building core-setup fails on Fedora 36 (which uses clang 14) because core-setup enables `-Werror` and then has code that produces warnings.

We should either fix the warnings, or disable `-Werror`.

The issue affects other Linux distributions as well, including Alpine: https://gitlab.alpinelinux.org/alpine/aports/-/tree/master/testing/dotnet31-build (see [core-setup_clang13-no-werror.patch](https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/testing/dotnet31-build/core-setup_clang13-no-werror.patch))

**Comment by NikolaMilosavljevic** [comment]
@vitek-karas this is host code - can you take a look?

**Comment by vitek-karas** [comment]
Actually @janvorli is a much better for this type of change.

One question - the linked change in the runtime https://github.com/dotnet/runtime/pull/63314 looks different - is there a reason for this difference?

**Comment by omajid** [comment]
@vitek-karas There's a few individual fixes needed to get all of .NET Core 3.1 building under clang 13.

This fix, which updates the [`SafeInt3.hpp`](https://github.com/dotnet/core-setup/pull/9171/files#diff-105332a39629ddf29fd120d6e77f918591e1f8bc7c221ece592ace570b3d5b39) file, doesn't apply in runtime. As far as I can tell, there's no SafeInt3.hpp in runtime at all.

The rest of the fixes in https://github.com/dotnet/runtime/pull/63314 take care of the changes that would be needed in ~~core-setup~~ coreclr and corefx for 3.1.

**Comment by omajid** [comment]
Any updates on this? Is this okay to merge?

**Comment by jeffschwMSFT** [comment]
we need to take for approval.  I will mark it as such.

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  We will take for consideration in 3.1.x

**Comment by omajid** [comment]
Thanks!

**Pull request merged**
